component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -1658,4 +1669,114 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
+                    SMSOTPConstants.X_FORWARDED_FOR).split(",")[0]);
+        } else {
+            eventProperties.put(IdentityEventConstants.EventProperty.CLIENT_IP, request.getRemoteAddr());
+        }
+
+        Event postOtpGenEvent = new Event(IdentityEventConstants.Event.POST_GENERATE_SMS_OTP, eventProperties);
+        try {
+            SMSOTPServiceDataHolder.getInstance().getIdentityEventService().handleEvent(postOtpGenEvent);
+        } catch (IdentityEventException e) {
+            String errorMsg = "An error occurred while triggering the event. " + e.getMessage();
+            throw new AuthenticationFailedException(errorMsg, e);
+        }
+    }
+
+    private void publishPostSMSOTPValidatedEvent(HttpServletRequest request,
+                                                 AuthenticationContext context) throws AuthenticationFailedException {
+
+        Map<String, Object> eventProperties = new HashMap<>();
+        AuthenticatedUser authenticatedUser = (AuthenticatedUser) context.getProperty(SMSOTPConstants
+                .AUTHENTICATED_USER);
+        eventProperties.put(IdentityEventConstants.EventProperty.CORRELATION_ID, context.getCallerSessionKey());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_NAME, authenticatedUser.getUserName());
+        eventProperties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, context.getTenantDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN, authenticatedUser
+                .getUserStoreDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.APPLICATION_NAME, context.getServiceProviderName());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_AGENT, request.getHeader(
+                SMSOTPConstants.USER_AGENT));
+
+        if (request.getHeader(SMSOTPConstants.X_FORWARDED_FOR) != null) {
+            eventProperties.put(IdentityEventConstants.EventProperty.CLIENT_IP, request.getHeader(
+                    SMSOTPConstants.X_FORWARDED_FOR).split(",")[0]);
+        } else {
+            eventProperties.put(IdentityEventConstants.EventProperty.CLIENT_IP, request.getRemoteAddr());
+        }
+
+        eventProperties.put(IdentityEventConstants.EventProperty.GENERATED_OTP, context.getProperty(
+                SMSOTPConstants.OTP_TOKEN));
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_INPUT_OTP, request.getParameter(
+                SMSOTPConstants.CODE));
+        eventProperties.put(IdentityEventConstants.EventProperty.OTP_USED_TIME, System.currentTimeMillis());
+
+        long otpGeneratedTime = (long) context.getProperty(SMSOTPConstants.SENT_OTP_TOKEN_TIME);
+        eventProperties.put(IdentityEventConstants.EventProperty.OTP_GENERATED_TIME,
+                otpGeneratedTime);
+
+        String otpExpiryTime = SMSOTPUtils.getTokenExpiryTime(context);
+        if (otpExpiryTime != null) {
```

Comment:
Use StringUtils.isNotEmpty()for null check and check this throughout the PR

Reply:
Fixed in 7d65437416ff6a2f9c995097fb0958fc95a58cf3